### PR TITLE
feat: add ocm app

### PR DESCRIPTION
**Pull request by JammingBen**
## Description
Implements federated sharing between multiple oCIS instances via [OpenCloudMesh (ocm)](https://oc.owncloud.com/opencloudmesh.html). Key features of this PR:

* New OCM app which is available via the app switcher menu (currently called `ScienceMesh`). There, users can see their existing connections, create new ones and accept invitations from other instances.
* Allows users to choose if they want to share a file with internal or external (= federated) users (cern-only; in our case the server always searches for all available users, no matter if internal or external).
* New action `Open remotely` lets users open federated shares on their original oC instance. We re-use most of the functionality for resolving public links for this. Auth handling works via a token coming from the backend.
* New filters on the "Shared with me" and "Shared with others" pages to filter by share types

Note that this implementation is not yet productized and is therefore missing some more things. The main goal here is to get this working for CERN. See https://github.com/owncloud/web/issues/9852#issuecomment-1808174243 for more details.

## Related Issue

- Fixes https://github.com/owncloud/web/issues/9268

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Follow ups

See https://github.com/owncloud/web/issues/9852#issuecomment-1808174243

### CODE REVIEW

File: packages/web-app-files/src/views/shares/SharedWithMe.vue [hunk 1]
```diff
@@ -19,21 +19,20 @@
...
             :items="shareTypes"
             :option-filter-label="$gettext('Filter share types')"
             :show-option-filter="true"
             id-attribute="key"
-            class="oc-mr-s"
+            class="oc-mx-s"
             display-name-attribute="label"
             filter-name="shareType"
           >
             <template #image="{ item }">
```
Comment (dschmidt) [inline packages/web-app-files/src/views/shares/SharedWithMe.vue:34]: We wanted to switch to another spot...

File: packages/web-app-files/src/views/shares/SharedWithMe.vue [hunk 2]
```diff
@@ -172,7 +171,7 @@ export default defineComponent({
     }
 
     const shareTypes = computed(() => {
-      const uniqueShareTypes = uniq(unref(currentItems).map((i) => i.share?.shareType))
+      const uniqueShareTypes = uniq(unref(storeItems).map((i) => i.share?.shareType))
```
Comment (dschmidt) [inline packages/web-app-files/src/views/shares/SharedWithMe.vue:174]: Why? This is more stable and recomputes less, but then it shows share types for potentially hidden shares, no?

Comment (dschmidt) [reply packages/web-app-files/src/views/shares/SharedWithMe.vue:174]: Ah, ok, lesser evil then. Thanks for clarification

File: packages/web-app-files/src/components/SideBar/Shares/Collaborators/InviteCollaborator/InviteCollaboratorForm.vue [hunk 3]
```diff
@@ -136,6 +156,7 @@ import {
 } from '@ownclouders/web-pkg'
 import { DateTime } from 'luxon'
 import { OcDrop } from 'design-system/src/components'
+import { FederatedConnection, FederatedUser } from '../../../../../../../web-app-ocm/src/types'
```
Comment (dschmidt) [inline packages/web-app-files/src/components/SideBar/Shares/Collaborators/InviteCollaborator/InviteCollaboratorForm.vue:159]: 🙈

File: dev/docker/ocis.web.config.json [hunk 4]
```diff
@@ -35,6 +35,7 @@
     "search",
     "external",
     "admin-settings",
+    "ocm",
```
Comment (dschmidt) [inline dev/docker/ocis.web.config.json:38]: should be opt in

File: packages/web-app-ocm/src/views/IncomingInvitations.vue [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+          <oc-select
+            v-model="provider"
+            :label="$gettext('Select institution of inviter')"
+            :options="providers"
+            class="oc-mb-m"
+            :loading="loading"
+          >
+            <template #option="{ full_name, domain }">
+              <span class="option">
+                <strong v-text="full_name" />
```
Comment (dschmidt) [inline packages/web-app-ocm/src/views/IncomingInvitations.vue:30]: What is this?

File: packages/web-client/src/helpers/space/functions.ts [hunk 6]
```diff
@@ -60,24 +78,33 @@ export function buildPublicSpaceResource(data): PublicSpaceResource {
...
   shareId: string | number
   shareName: string
   serverUrl: string
 }): ShareSpaceResource {
+  let id
+  if (driveAliasPrefix === 'ocm-share') {
+    id = `${OCM_PROVIDER_ID}$${shareId}!${shareId}`
+  } else {
+    id = [SHARE_JAIL_ID, shareId].join('!')
+  }
```
Comment (dschmidt) [inline packages/web-client/src/helpers/space/functions.ts:96]: Afaict this is the only place where we need to distinguish between regular `share` and `ocm-share`  in the drive alias -  do we want to introduce this new drive alias prefix just for this?

An alternative would be a query param or find another clever solution to delay using the id until the actual share is loaded...

File: dev/docker/ocis.web.config.json (hunk 4 above)
Comment (dschmidt) [reply dev/docker/ocis.web.config.json:38]: this is just for the dev stack, let's keep it

File: packages/web-app-ocm/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
+import App from './views/App.vue'
+import { defineWebApplication, useRouter } from '@ownclouders/web-pkg'
+import translations from '../l10n/translations.json'
+import { extensions } from './extensions'
+import { RouteRecordRaw } from 'vue-router'
+import { useGettext } from 'vue3-gettext'
+
+const appInfo = {
+  name: 'ScienceMesh',
+  id: 'sciencemesh-app',
```
Comment (dschmidt) [inline packages/web-app-ocm/src/index.ts:10]: What name/id should we use?

File: packages/web-app-files/src/components/SideBar/Shares/Collaborators/InviteCollaborator/InviteCollaboratorForm.vue [hunk 8]
```diff
@@ -366,6 +398,13 @@ export default defineComponent({
 
       this.searchInProgress = true
 
+      // CERN
+      if (this.isRunningOnEos) {
+        query = `${
+          this.accountTypes.find((t) => t.description === this.accountType)?.prefix || ''
```
Comment (lookacat) [inline packages/web-app-files/src/components/SideBar/Shares/Collaborators/InviteCollaborator/InviteCollaboratorForm.vue:404]: for readability reasons please extract this into a var :)

File: packages/web-app-files/src/views/shares/SharedWithMe.vue [hunk 9]
```diff
@@ -8,16 +8,34 @@
...
+            :items="shareTypes"
+            :option-filter-label="$gettext('Filter share types')"
+            :show-option-filter="true"
+            id-attribute="key"
+            class="share-type-filter oc-mx-s"
+            display-name-attribute="label"
+            filter-name="shareType"
+          >
+            <template #item="{ item }">
+              <span class="oc-ml-s">{{ item.label }}</span>
```
Comment (lookacat) [inline packages/web-app-files/src/views/shares/SharedWithMe.vue:36]: can be shortend with v-text

File: packages/web-app-files/src/views/shares/SharedWithOthers.vue [hunk 10]
```diff
@@ -8,6 +8,28 @@
...
+            :items="shareTypes"
+            :option-filter-label="$gettext('Filter share types')"
+            :show-option-filter="true"
+            id-attribute="key"
+            class="share-type-filter oc-mx-s"
+            display-name-attribute="label"
+            filter-name="shareType"
+          >
+            <template #item="{ item }">
+              <span class="oc-ml-s">{{ item.label }}</span>
```
Comment (lookacat) [inline packages/web-app-files/src/views/shares/SharedWithOthers.vue:29]: can also be shorten with v-text :)

File: packages/web-app-ocm/src/views/App.vue [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+          store.dispatch('showMessage', {
+            title: $gettext('New federated connection'),
+            status: 'success',
+            desc: $gettext('You can share with and recieve shares from %{user} now', {
+              user: unref(highlightedConnections)[0].display_name
+            })
+          })
+        } else if (unref(highlightedConnections).length > 1) {
+          let newConnections
+          unref(highlightedConnections).forEach((c) => (newConnections += c.display_name + ', '))
```
Comment (lookacat) [inline packages/web-app-ocm/src/views/App.vue:80]: maybe this works better:
```
let newConnections = unref(highlightedConnections)
                     .map(c => c.display_name)
                     .join(', ');
```

File: packages/web-app-ocm/src/views/ConnectionsPanel.vue [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+      try {
+        await clientService.httpAuthenticated.delete('/sciencemesh/delete-accepted-user', {
+          params: {
+            user_id: user.user_id,
+            idp: user.idp
+          }
+        })
+
+        const updatedConnections = []
+        props.connections.forEach((connection) => {
```
Comment (lookacat) [inline packages/web-app-ocm/src/views/ConnectionsPanel.vue:164]: Improvement idea:
```
const updatedConnections = props.connections.filter(connection =>
    connection.user_id !== user.user_id && connection.idp !== user.idp);
```

File: packages/web-app-ocm/src/views/IncomingInvitations.vue [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+      acceptInvitationButtonDisabled,
+      acceptInvite
+    }
+  },
+  methods: { $gettext }
+})
+</script>
+
+<style lang="scss" scoped>
+.sciencemesh-app {
```
Comment (lookacat) [inline packages/web-app-ocm/src/views/IncomingInvitations.vue:203]: empty class can be removed?

**Review by lookacat — APPROVED** [review]
Some improvement ideas, but I can only approve it code wise, i just trust that it works :)

File: packages/web-app-ocm/l10n/translations.json [hunk 14]
```diff
@@ -0,0 +1 @@
+{}
```
[the remark this answers is not in the document]
Comment (dschmidt) [reply packages/web-app-ocm/l10n/translations.json:1]: Yeah, handled by transifex

File: packages/web-app-files/tests/unit/views/shares/SharedWithOthers.spec.ts [hunk 15]
```diff
@@ -44,6 +52,25 @@ describe('SharedWithOthers view', () => {
       )
     })
   })
+  describe('filter', () => {
+    describe('share type', () => {
+      it('shows filter if more than one share types are present', () => {
```
Comment (kulmann) [inline packages/web-app-files/tests/unit/views/shares/SharedWithOthers.spec.ts:57]:
```suggestion
      it('shows filter if multiple share types are present', () => {
```

File: packages/web-app-importer/src/extensions.ts [hunk 16]
```diff
@@ -112,6 +112,7 @@ export const extensions = ({ applicationConfig }: ApplicationSetupOptions) => {
         {
           id: 'com.github.owncloud.web.import-file',
           type: 'action',
+          scopes: ['files', 'upload-menu'],
```
Comment (kulmann) [inline packages/web-app-importer/src/extensions.ts:115]: do we already have any semantics for the scopes? 🙈 is `files` referring to `resources` or `files app`?

File: packages/web-app-ocm/src/index.ts [hunk 17]
```diff
@@ -0,0 +1,64 @@
+import App from './views/App.vue'
+import { defineWebApplication, useRouter } from '@ownclouders/web-pkg'
+import translations from '../l10n/translations.json'
+import { extensions } from './extensions'
+import { RouteRecordRaw } from 'vue-router'
+import { useGettext } from 'vue3-gettext'
+
+const appInfo = {
+  name: 'ScienceMesh',
```
Comment (kulmann) [inline packages/web-app-ocm/src/index.ts:9]: please wrap the name in a `$gettext` dummy or move it into the `defineWebApplication` body.

File: packages/web-app-ocm/src/index.ts (hunk 7 above)
Comment (kulmann) [reply packages/web-app-ocm/src/index.ts:10]: why not just `ocm` as id?

File: packages/web-app-ocm/src/views/App.vue [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+          newConnections.slice(0, -2)
```
Comment (kulmann) [inline packages/web-app-ocm/src/views/App.vue:81]: `slice` doesn't modify the array, so this statement doesn't have any effect.

File: packages/web-app-ocm/src/views/App.vue (hunk 11 above)
Comment (kulmann) [reply packages/web-app-ocm/src/views/App.vue:80]: Yes, and make it a `const`

File: packages/web-app-ocm/src/views/App.vue [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+          store.dispatch('showMessage', {
+            title: $gettext('New federated connections'),
+            status: 'success',
+            desc: $gettext('You can share with and recieve shares from %{newConnections} now', {
```
Comment (kulmann) [inline packages/web-app-ocm/src/views/App.vue:85]:
```suggestion
            desc: $gettext('You can share with and receive shares from %{ connections } now', {
```

File: packages/web-app-ocm/src/views/IncomingInvitations.vue [hunk 20]
```diff
@@ -0,0 +1,208 @@
...
+            <template #option="{ full_name, domain }">
+              <span class="option">
+                <strong v-text="full_name" />
+              </span>
+              <span class="option" v-text="domain" />
+            </template>
+            <template #no-options> No institutions found with this name </template>
+            <template #selected-option="{ full_name, domain }">
+              <strong class="oc-mr-s" v-text="full_name" />
+              <small v-text="domain.length > 17 ? domain.slice(0, 20) + '...' : domain" />
```
Comment (kulmann) [inline packages/web-app-ocm/src/views/IncomingInvitations.vue:37]: could you add the full domain as a tooltip?

File: packages/web-app-ocm/src/views/IncomingInvitations.vue [hunk 21]
```diff
@@ -0,0 +1,208 @@
...
+    const token = ref(undefined)
+    const provider = ref(undefined)
+    const providers = ref([])
+    const loading = ref(true)
+    const providerError = ref(false)
+    const tokenInput = ref<VNodeRef>()
+
+    const helperContent = computed(() => {
+      return {
+        text: $gettext('Once you accept invitation, the inviter will be added to your connections.')
```
Comment (kulmann) [inline packages/web-app-ocm/src/views/IncomingInvitations.vue:87]:
```suggestion
        text: $gettext('Once you accept the invitation, the inviter will be added to your connections.')
```

File: packages/web-app-ocm/src/views/IncomingInvitations.vue [hunk 22]
```diff
@@ -0,0 +1,208 @@
...
+      }
+    })
+
+    const acceptInvitationButtonDisabled = computed(() => {
+      return !unref(token) || !unref(provider) || unref(provider).full_name === 'Unknown provider'
+    })
+
+    const providerErrorMessage = computed(() => {
+      return isMyProviderSelectedProvider(unref(provider))
+        ? $gettext('You cannot select your institution')
```
Comment (kulmann) [inline packages/web-app-ocm/src/views/IncomingInvitations.vue:97]:
```suggestion
        ? $gettext('You cannot select your own institution')
```

File: packages/web-app-ocm/src/views/OutgoingInvitations.vue [hunk 23]
```diff
@@ -0,0 +1,351 @@
...
+      <app-loading-spinner v-if="loading" />
+      <template v-else>
+        <no-content-message
+          v-if="!sortedTokens.length"
+          id="invite-tokens-empty"
+          class="files-empty"
+          icon="user-shared"
+        >
+          <template #message>
+            <span v-text="$gettext('You have no valid invitation links')" />
```
Comment (kulmann) [inline packages/web-app-ocm/src/views/OutgoingInvitations.vue:72]:
```suggestion
            <span v-text="$gettext('You have no invitation links')" />
```

File: packages/web-app-ocm/src/views/OutgoingInvitations.vue [hunk 24]
```diff
@@ -0,0 +1,351 @@
...
+        text: $gettext(
+          'Create an invitation link and send it to the person you want to share with.'
+        )
+      }
+    })
+
+    const generateToken = async () => {
+      const { description, recipient } = unref(formInput)
+      if (recipient.length > 0 && !EmailValidator.validate(recipient)) {
+        emailErrorMessage.value = $gettext('Please enter a valid email adress!')
```
Comment (kulmann) [inline packages/web-app-ocm/src/views/OutgoingInvitations.vue:190]:
```suggestion
        emailErrorMessage.value = $gettext('Please enter a valid email address!')
```

File: packages/web-app-ocm/src/views/OutgoingInvitations.vue [hunk 25]
```diff
@@ -0,0 +1,351 @@
...
+      store.dispatch('showMessage', {
+        title: $gettext('Invition link copied'),
+        desc: $gettext('Invitation link has been copied to your clipboard.')
+      })
+    }
+    const copyToken = (rowData) => {
+      navigator.clipboard.writeText(rowData.item.token)
+      store.dispatch('showMessage', {
+        title: $gettext('Invite token copied'),
+        desc: $gettext('Invitate token has been copied to your clipboard.')
```
Comment (kulmann) [inline packages/web-app-ocm/src/views/OutgoingInvitations.vue:282]:
```suggestion
        desc: $gettext('Invite token has been copied to your clipboard.')
```

File: packages/web-app-ocm/src/views/OutgoingInvitations.vue [hunk 26]
```diff
@@ -0,0 +1,351 @@
...
+      })
+    }
+    const errorPopup = (error: Error) => {
+      console.error(error)
+      store.dispatch('showErrorMessage', {
+        title: $gettext('Error'),
+        desc: $gettext('An error occurred by generating the token'),
```
Comment (kulmann) [inline packages/web-app-ocm/src/views/OutgoingInvitations.vue:289]:
```suggestion
        desc: $gettext('An error occurred when generating the token'),
```

File: packages/web-app-files/src/components/SideBar/Shares/Collaborators/InviteCollaborator/InviteCollaboratorForm.vue [hunk 27]
```diff
@@ -201,6 +222,29 @@ export default defineComponent({
       displayPositionedDropdown(dropdown.tippy, event, unref(contextMenuButtonRef))
     }
 
+    const federatedUsers = ref([] as FederatedUser[])
+    onMounted(async () => {
+      // // HACK: remove when federated users are returned from search
+      // try {
```
[the remark this answers is not in the document]
Comment (dschmidt) [reply packages/web-app-files/src/components/SideBar/Shares/Collaborators/InviteCollaborator/InviteCollaboratorForm.vue:228]: Yes, we had already removed it but then couldn't test anymore so we reverted the removal. Obviously they are supposed to go before merge.

File: packages/web-app-ocm/src/views/IncomingInvitations.vue (hunk 22 above)
Comment (dschmidt) [reply packages/web-app-ocm/src/views/IncomingInvitations.vue:97]: Thinking about it ... why would it be in the list then at all?!
Let's remove it from the list instead...

File: packages/web-app-files/src/services/folder/loaderSpace.ts [hunk 28]
```diff
@@ -84,6 +84,11 @@ export class FolderLoaderSpace implements FolderLoader {
           }
         }
 
+        // FIXME: HACK:
```
[the remark this answers is not in the document]
Comment (JammingBen) [reply packages/web-app-files/src/services/folder/loaderSpace.ts:87]: FIXME for later, will add a comment though.

File: packages/web-app-importer/src/extensions.ts (hunk 16 above)
Comment (JammingBen) [reply packages/web-app-importer/src/extensions.ts:115]: Not yet 😅

> is `files` referring to `resources` or `files app`?

Hard to say tbh. Both, kinda 😄

File: packages/web-client/src/webdav/client/builders.ts [hunk 29]
```diff
@@ -41,6 +41,11 @@ export const buildPublicLinkAuthHeader = (password: string) => {
 
 export const buildAuthHeader = (token: string, space: SpaceResource = null): Headers => {
   if (isPublicSpaceResource(space)) {
+    // TODO: make check cleaner
+    if (space.driveAlias.startsWith('ocm/')) {
+      return { Authorization: `Bearer ${space.id}` }
```
Comment (kulmann) [inline packages/web-client/src/webdav/client/builders.ts:46]: eh... what?! space id is the bearer token?!

File: packages/web-pkg/src/components/FilesList/ContextActions.vue [hunk 30]
```diff
@@ -75,6 +77,14 @@ export default defineComponent({
     const { actions: showSharesActions } = useFileActionsShowShares({ store })
     const { actions: openShortcutActions } = useFileActionsOpenShortcut({ store })
 
+    const extensionRegistry = useExtensionRegistry()
+    const extensionContextActions = computed(() => {
+      return extensionRegistry
+        .requestExtensions<ActionExtension>('action')
+        .filter(({ scopes }) => scopes.includes('files.context-menu'))
```
Comment (kulmann) [inline packages/web-pkg/src/components/FilesList/ContextActions.vue:84]: I think I only saw `context-menu` defined as a scope, not  `files.context-menu`. How is this supposed to work?

File: packages/web-runtime/src/services/auth/publicLinkManager.ts [hunk 31]
```diff
@@ -40,6 +40,14 @@ export class PublicLinkManager {
     sessionStorage.setItem(PublicLinkManager.buildStorageKey(token, 'resolved'), resolved + '')
   }
 
+  setType(token: string, type: 'ocm' | 'public-link'): void {
```
Comment (kulmann) [inline packages/web-runtime/src/services/auth/publicLinkManager.ts:43]: Saw `'ocm' | 'public-link'` quite a few times now... maybe makes sense to introduce a `LinkShareType` type or something similar...

File: packages/web-client/src/webdav/client/builders.ts (hunk 29 above)
Comment (dschmidt) [reply packages/web-client/src/webdav/client/builders.ts:46]: Yes ... ... ... ... ... ... ... .. .. ..

Comment (dschmidt) [reply packages/web-client/src/webdav/client/builders.ts:46]: the ocm token *must* be provided as bearer token (although it's laready part of the request url ...)

Comment (kulmann) [reply packages/web-client/src/webdav/client/builders.ts:46]: That was the detail that I was still missing... ok, thanks for explaining. This is so horrible...

File: packages/web-app-files/src/components/SideBar/Shares/Collaborators/InviteCollaborator/InviteCollaboratorForm.vue (hunk 27 above)
[the remark this answers is not in the document]
Comment (dschmidt) [reply packages/web-app-files/src/components/SideBar/Shares/Collaborators/InviteCollaborator/InviteCollaboratorForm.vue:228]: @aduffeck will fix the search for us, but there are some more important blockers right now

**Review by kulmann — APPROVED** [review]
Let's merge and iterate... thanks to both of you @JammingBen @dschmidt for going down this rabbit hole! :heart:

**Comment by dschmidt** [comment]
🎉 🎉 🎉 🎉 🎉

**Pull request merged**